### DISCUSSION TITLE
pip-compile can fail artifact update check with multiple inputs

### DISCUSSION THREAD

**Discussion by not7cd** [renovatebot/renovate#29656] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

CircleCI

## Problem

Renovate will fail a check for artifact update. Even though it eventually succeeds in this update by running `pip-compile` more than once.

![image](https://github.com/renovatebot/renovate/assets/4332632/0099b93d-701d-4fe2-8941-a328a8fbf372)

## Description
Given this requirement file hierarchy:

```mermaid

       graph LR
         requirements-jupyter.txt[[requirements-jupyter.txt]]
         requirements.txt[[requirements.txt]]
         irrelevant.txt[[irrelevant.txt]]
         requirements-shared.in --> irrelevant.txt
         requirements-jupyter.in --> requirements-jupyter.txt
         requirements.txt --> requirements-jupyter.in
         requirements-shared.in --> requirements.txt
         requirements.in --> requirements.txt 
         
 ```
 
We can expect Renovate to run pip-compile more than once on a lock file. This is because all descending lock files are added to a package file, as introduced in https://github.com/renovatebot/renovate/pull/28052. 

This indeed happens, but in an order that throws **artifact update error** for requirements-jupyter.txt.

```log

DEBUG: updateArtifacts for nonUpdatedPackageFiles
# ...
DEBUG: pipCompile.updateArtifacts(requirements-jupyter.in->["requirements-jupyter.txt"]) 
DEBUG: pip-compile: Failed to run command
       "err": {
         "cmd": "/bin/sh -c pip-compile --all-extras --no-emit-index-url --output-file=requirements-jupyter.txt requirements-jupyter.in --upgrade-package=foo==1.2.3",
         "stderr": "  ERROR: Cannot install foo==1.2.3 because these package versions have conflicting dependencies. ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts",
         "stdout": "",
         "options": {
             ...
         },
         "exitCode": 1,
         "name": "ExecError",
         "message": "..."
       }
DEBUG: pipCompile.updateArtifacts(requirements-shared.in->["irrelevant.txt","requirements.txt","requirements-jupyter.txt"]) 
DEBUG: pipCompile.updateArtifacts(requirements.in->["requirements.txt","requirements-jupyter.txt"]) 
# ...
```

The `pipCompile.updateArtifacts(requirements-jupyter.in->["requirements-jupyter.txt"])` will fail because it depends on `requirements.txt` which is not yet compiled. It also all happens within the context of `updateArtifacts for nonUpdatedPackageFiles`

## Possible fix

Keep the package file order based on topological sort during `artifactUpdate`. I don't remember if this order was ever respected. 

Alternatively, remove `artifactErrors` if it succeds in another `artifactUpdate` execution. This must happen outside `pip-compile` module.

The fix may require looking into the whole `updateArtifacts` section in get-updated. https://github.com/renovatebot/renovate/blob/7178da30743b88b289d2f2cd859891d4c07be3ab/lib/workers/repository/update/branch/get-updated.ts#L298-L394

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by not7cd** [comment]
@mbudnek, did you encounter similar glitches? I think we discussed something, but I'm not sure if there was a conclusion.

**Reply by mbudnek** [reply]
Yes, I have https://github.com/renovatebot/renovate/pull/28671 open to fix this, but it's been pending for a while now awaiting review.

**Reply by viceice** [reply]
I didn't had the time to review carefully yet. sorry.
